cmd/default-domain/main.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/knative/pkg/logging"
+	"github.com/knative/pkg/system"
+	cicfg "github.com/knative/serving/pkg/reconciler/v1alpha1/clusteringress/config"
+	routecfg "github.com/knative/serving/pkg/reconciler/v1alpha1/route/config"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	masterURL  = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	kubeconfig = flag.String("kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	magicDNS   = flag.String("magic-dns", "", "The hostname for the magic DNS service, e.g. xip.io or nip.io")
+)
+
+const (
+	// Interval to poll for objects.
+	pollInterval = 10 * time.Second
+	// How long to wait for objects.
+	waitTimeout = 20 * time.Minute
+	appName     = "default-domain"
+)
+
+func kubeClientFromFlags() (*kubernetes.Clientset, error) {
+	cfg, err := clientcmd.BuildConfigFromFlags(*masterURL, *kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("Error building kubeconfig: %v", err)
+	}
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("Error building kube clientset: %v", err)
+	}
+	return kubeClient, nil
+}
+
+func lookupConfigMap(kubeClient *kubernetes.Clientset, name string) (*corev1.ConfigMap, error) {
+	return kubeClient.CoreV1().ConfigMaps(system.Namespace()).Get(name, metav1.GetOptions{})
+}
+
+func lookupIngressGateway(kubeClient *kubernetes.Clientset) (*corev1.Service, error) {
+	// Fetch and parse the Istio-ClusterIngress ConfigMap from the system namespace.
+	istioCM, err := lookupConfigMap(kubeClient, cicfg.IstioConfigName)
+	if err != nil {
+		return nil, err
+	}
+	istioConfig, err := cicfg.NewIstioFromConfigMap(istioCM)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing ConfigMap: %v", err)
+	}
+	// Parse the service name to determine which Kubernetes Service resource to fetch.
+	serviceURL := istioConfig.IngressGateways[0].ServiceURL
+	parts := strings.SplitN(serviceURL, ".", 3)
```

Comment:
Could you please add a comment for the format `serviceURL`? It would make it easy to understand why it is `3` here.

Reply:
Done